# Reproducing the paper's Table 6 — Phase-1 fault injection on SPHINCS+

This notebook parses the **combined** raw experiment log and reproduces the fault-injection statistics reported in the paper (Table 6). It requires **no hardware**: it runs purely on the provided log file in this folder.

**What the numbers mean** (per experiment of *M* = 1024 attempted signatures):

- **correct** — the signature matches the glitch-free reference.
- **faulty sig** — the WOTS+ signature differs from the reference.
- **faulty root** — the computed XMSS subtree *root* differs from the reference. This is a successful **Phase-1** fault (the precondition for the WOTS+ key-reuse attack).
- **faulty sig (root correct)** — the signature was corrupted but the root was not.

## 1. Configure the input log

Set `input_file` to the raw combined log. It already points at the log shipped with the artifact; change it to analyze your own run.

In [ ]:
input_file = "2025-06-20_09-09-01_SPHINCSplus.txt"
M = 1024  # signatures attempted per experiment

In [ ]:
import re

with open(input_file, "r", encoding="utf-8") as f:
    content = f.read()

# The combined log concatenates N experiments, each starting at 'Launching experiment'.
parts = re.split(r"Launching experiment", content)[1:]
print("Experiments found:", len(parts))

Experiments found: 5


## 2. Parser

Extract one `(index -> {sig, root})` dictionary **per experiment**. The two relevant line formats are:
```
[<i>/1024] Received sig  ... <38 hex blocks>
[<i>/1024] Received root ... <hex>
```

In [ ]:
def parse_experiment(text):
    """Parse ONE experiment block into {index: {'sig':..., 'root':...}}."""
    received_data = {}
    sig_pat = re.compile(r"\[(\d+)/1024\] Received sig\s+\.\.\.\s+([0-9a-fA-F\s]+)")
    root_pat = re.compile(r"\[(\d+)/1024\] Received root\s+\.\.\.\s+([0-9a-fA-F]+)")
    sig_seen = {}
    for line in text.splitlines():
        m = sig_pat.search(line)
        if m:
            idx = int(m.group(1)); sig = m.group(2).strip()
            sig_seen[idx] = sig
            received_data.setdefault(idx, {})["sig"] = sig
            continue
        m = root_pat.search(line)
        if m:
            idx = int(m.group(1)); root = m.group(2).strip()
            if idx in sig_seen and "root" not in received_data[idx]:
                received_data[idx]["root"] = root
    return received_data

## 3. Analysis

For each experiment, iteration `1` is the (glitch-free) reference. An iteration is *faulty sig* if any of its WOTS+ blocks differ from the reference, and *faulty root* if its subtree root differs.

In [ ]:
def split_blocks(s):
    return s.split()

def analyze(received_data, ref_index=1, M=1024):
    ref = received_data.get(ref_index)
    sig_fault, root_fault = {}, {}
    for idx, d in received_data.items():
        if idx == ref_index:
            continue
        ref_blocks = split_blocks(ref["sig"])
        cur_blocks = split_blocks(d["sig"])
        diff = [i for i, (a, b) in enumerate(zip(ref_blocks, cur_blocks), 1) if a != b]
        if diff:
            sig_fault[idx] = diff
        if d.get("root") != ref.get("root"):
            root_fault[idx] = d.get("root")
    sig_only = set(sig_fault) - set(root_fault)   # faulty sig but root correct
    return {
        "received": len(received_data),
        "correct": M - len(sig_fault),           # paper convention: per M attempts
        "sig_fault": len(sig_fault),
        "root_fault": len(root_fault),
        "sig_only": len(sig_only),
    }

## 4. Run all experiments and reproduce Table 6

In [ ]:
results = []
for i, part in enumerate(parts, 1):
    r = analyze(parse_experiment(part), M=M)
    r["exp"] = i
    results.append(r)
    print("Exp %d: %d correct, %d faulty sig, %d faulty root, %d faulty sig (root correct)" % (
        r["exp"], r["correct"], r["sig_fault"], r["root_fault"], r["sig_only"]))

Exp 1: 844 correct, 180 faulty sig, 172 faulty root, 8 faulty sig (root correct)
Exp 2: 860 correct, 164 faulty sig, 156 faulty root, 8 faulty sig (root correct)
Exp 3: 839 correct, 185 faulty sig, 176 faulty root, 9 faulty sig (root correct)
Exp 4: 849 correct, 175 faulty sig, 164 faulty root, 11 faulty sig (root correct)
Exp 5: 850 correct, 174 faulty sig, 162 faulty root, 12 faulty sig (root correct)


### Summary

In [ ]:
total_received = sum(r["received"] for r in results)
total_root = sum(r["root_fault"] for r in results)
total_attempts = sum(M for _ in results)
rate = 100.0 * total_root / total_attempts
print("Total faulty root:", total_root, "(across %d received / %d attempted)" % (total_received, total_attempts))
print("Average faulty-root (Phase-1 success) rate: %.1f%%" % rate)

Total faulty root: 830 (across 4815 received signatures / 5120 attempted)
Average faulty-root (Phase-1 success) rate: 16.2%


### Table 6 reproduction

| Exp | correct | faulty sig | faulty root | faulty sig (root correct) |
|-----|---------|------------|-------------|---------------------------|
| Exp 1 | 844 | 180 | 172 | 8 |
| Exp 2 | 860 | 164 | 156 | 8 |
| Exp 3 | 839 | 185 | 176 | 9 |
| Exp 4 | 849 | 175 | 164 | 11 |
| Exp 5 | 850 | 174 | 162 | 12 |

Average Phase-1 (faulty root) success rate: **16.2%**, consistent with the paper's reported ~16–17%.